packages/navigator/src/browser/search-box.ts

```diff
@@ -0,0 +1,271 @@
+/*
+ * Copyright (C) 2018 TypeFox and others.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+import { inject, injectable } from 'inversify';
+import { KeyCode, Key } from '@theia/core/lib/browser';
+import { BaseWidget } from '@theia/core/lib/browser/widgets/widget';
+import { Event, Emitter } from '@theia/core/lib/common/event';
+import { DisposableCollection } from '@theia/core/lib/common/disposable';
+import { FileNavigatorSearch } from './navigator-search';
+
+export namespace SearchBox {
+
+    /**
+     * Search box widget factory.
+     */
+    export const Factory = Symbol('SearchBox.Factory');
+    export interface Factory {
+
+        /**
+         * Creates a new search box widget with the given initializer properties.
+         */
+        (props: SearchBox.Props): SearchBox.Widget;
+
+    }
+
+    /**
+     * Initializer properties for the search box widget.
+     */
+    @injectable()
+    export class Props {
+
+        /**
+         * Throttle delay (in milliseconds) that is used before notifying clients about search data updates.
+         */
+        readonly delay: number;
+
+    }
+
+    export namespace Props {
+
+        /**
+         * The default search box widget option.
+         */
+        export const DEFAULT: SearchBox.Props = {
+            delay: 50
+        };
+
+    }
+
+    /**
+     * The search box widget.
+     */
+    @injectable()
+    export class Widget extends BaseWidget {
```

Comment:
not sure, why `SearchBoxWidget` cannot be the top-level element, it seems to be the main concept here. And `Factory`, `Props` belong `SearchBoxWidget` namespace

Reply:
I completely agree with you, damn symbols and namespaces and classes case the problem. For instance if I have the following:

```typescript
class A {
  constructor(p: A.Props) {}
}

namespace A {
  export class Props {}
}

```

Then it does not work with inversify. And I cannot have the namespace before the class. Sometimes, we have symbols too which adds one additional complexity.

Reply:
I think the snippet which you posted should work. Usually it is something like that:
```ts
const A = Symbol('A');
interface A {
}
namespace A { // 2 collided runtime variables with A name
}
```

Reply:
OK, let's try again then.